### PR TITLE
[BOT] Added implementation for Optimus Prime and its abilities

### PR DESCRIPTION
**Pull request by jbureau88**
Need to determine why triggers firing more than once but apart from that the rest works.

- Bolster is triggering twice with one attacker ? Why?
- Autobot leader convert trigger twice?
- Autobot leader trample given to modified creature trigger more than once?

Any pointers in understanding why additional triggers are happening would be appreciated!

### CODE REVIEW

File: Mage/src/main/java/mage/constants/SpellAbilityCastMode.java [hunk 1]
```diff
@@ -13,7 +13,8 @@ public enum SpellAbilityCastMode {
     MADNESS("Madness"),
     FLASHBACK("Flashback"),
     BESTOW("Bestow"),
-    DISTURB("Disturb");
+    DISTURB("Disturb"),
+    MORE_THAN_MEETS_THE_EYE("More Than Meets The Eye");
```
Comment (JayDi85) [inline Mage/src/main/java/mage/constants/SpellAbilityCastMode.java:17]: No card hacks allows in such game engine's deep level. Must be another way to implement your ability.

Comment (Grath) [reply Mage/src/main/java/mage/constants/SpellAbilityCastMode.java:17]: It's not a hack, it's correcting More Than Meets The Eye (already implemented that deep-game-engine level way by @theelk801 ) to be actually (correctly) distinct from Disturb which it was based on.

**Comment by JayDi85** [comment]
There are multiple ways to get duplicated events:
 * one of the popular bug: ZONE_CHANGE_GROUP problem, see #8009;
 * another popular and fixable bug: cards moving and processing one by one instead of group (search code for `game.getState().processAction(game);` for correct usage).

File: Mage/src/main/java/mage/constants/SpellAbilityCastMode.java (hunk 1 above)
Comment (jbureau88) [reply Mage/src/main/java/mage/constants/SpellAbilityCastMode.java:17]: As @Grath mentions, I was simply trying to separate from disturb copy paste remnants, I can revert that part if need be.

Comment (JayDi85) [reply Mage/src/main/java/mage/constants/SpellAbilityCastMode.java:17]: OK, I see, you can keep it. But must fix:
1. Add new type to `case DISTURB` code (search it);
2. Use `setCardName` with additional name like `+ " with More Than Meets the Eye"`

**Comment by jbureau88** [comment]
> There are multiple ways to get duplicated events:
> 
> * ~one of the popular~ (not anymore) bug: ZONE_CHANGE_GROUP problem, see [ZONE_CHANGE_GROUP event must be reworked to fix duplicated events #8009](https://github.com/magefree/mage/issues/8009);
> * another popular and fixable bug: cards moving and processing one by one instead of group (search code for `game.getState().processAction(game);` for correct usage).

Fixed now thanks

**Comment by jbureau88** [comment]
Oops meant to close the thread not the pull request..

File: Mage.Sets/src/mage/cards/o/OptimusPrimeHero.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            return false;
+        }
+
+        controller.moveCards(card, Zone.BATTLEFIELD, source, game);
+
+        game.getState().processAction(game);
+
+        Permanent permanent = game.getPermanent(source.getSourceId());
+        if (permanent.isTransformable() && !permanent.isTransformed()) {
+            permanent.transform(source, game);
```
[the remark this answers is not in the document]
Comment (jbureau88) [reply Mage.Sets/src/mage/cards/o/OptimusPrimeHero.java:85]: Ok I'll try to go back to the other way but it was causing the doubling of effects how it previously was.

[the remark this answers is not in the document]
Comment (jbureau88) [reply Mage.Sets/src/mage/cards/o/OptimusPrimeHero.java]: It works ! The `game.getState().processAction(game);` really did fix the issues 👍

File: Mage.Sets/src/mage/cards/o/OptimusPrimeAutobotLeader.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        // Trample
+        this.addAbility(TrampleAbility.getInstance());
+
+        this.getSpellAbility().addWatcher(new OptimusPrimeAutobotLeaderWatcher());
+
+        // Whenever you attack, bolster 2.
+        this.addAbility(new AttacksWithCreaturesTriggeredAbility(new BolsterEffect(2), 1));
+
+        // The chosen creature gains trample until end of turn.
+        this.addAbility(new AutobotLeaderAddTrampleTriggeredAbility());
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/o/OptimusPrimeAutobotLeader.java:53]: This is a hacky way to implement it and could trigger off unrelated counters being added. The watcher checking name is also brittle. I think what actually needs to happen is `BolsterEffect` gets an optional second parameter to take an additional `Effects` to add to the chosen creature. Then this card can add the gain trample effect, and the delayed triggered ability that stores a MageObjectReference to the source and converts it if possible.

**Review by xenohedron — CHANGES_REQUESTED** [review]

**Comment by xenohedron** [comment]
Alright I reimplemented this and wrote a test

**Pull request merged**